### PR TITLE
Fixes 1973: Add repository snapshot endpoint

### PR DESCRIPTION
**Pull request by Andrewgdewar**
## Summary

Adds a repository snapshot endpoint for manually triggering snapshots

## Testing steps

**Method 1**

Easiest method of testing would be to change the following lines in the FE > ContentApi.ts

Screenshot 2023-11-02 at 2 54 47 PM

And disable this one line in the ContentListTable

Screenshot 2023-11-02 at 2 55 52 PM

So that when you click the "Introspect Now" button in the kebab, it is actually requesting to sync a snapshot.

**Method 2**
Alternatively call the new endpoint directly: /repositories/{uuid}/snapshot/ [post] 
You'll need to have created a repository.

## Checklist

- [x] Tested with snapshotting feature disabled and pulp server URL not configured if appropriate

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-1973

File: pkg/handler/repositories.go [hunk 1]
```diff
@@ -52,6 +52,7 @@ func RegisterRepositoryRoutes(engine *echo.Group, daoReg *dao.DaoRegistry,
 	addRoute(engine, http.MethodPost, "/repositories/bulk_delete/", rh.bulkDeleteRepositories, rbac.RbacVerbWrite)
 	addRoute(engine, http.MethodPost, "/repositories/", rh.createRepository, rbac.RbacVerbWrite)
 	addRoute(engine, http.MethodPost, "/repositories/bulk_create/", rh.bulkCreateRepositories, rbac.RbacVerbWrite)
+	addRoute(engine, http.MethodPost, "/repositories/:uuid/snapshot/", rh.syncSnapshot, rbac.RbacVerbWrite)
```
Comment (jlsherrill) [inline pkg/handler/repositories.go:55]: what do you think about moving this to the snapshot  handler (and maybe renaming it 'create') ?

Comment (Andrewgdewar) [reply pkg/handler/repositories.go:55]: I think that makes sense.
When you say "this" you mean the method, the tests, all of it?  
Or just the route?

File: pkg/handler/repositories.go [hunk 2]
```diff
@@ -465,6 +466,59 @@ func (rh *RepositoryHandler) bulkDeleteRepositories(c echo.Context) error {
...
+	status := config.StatusPending
+
+	repoUpdate = dao.RepositoryUpdate{
+		UUID:   repo.UUID,
+		Status: &status,
+	}
+
+	if err := rh.DaoRegistry.Repository.Update(repoUpdate); err != nil {
+		return ce.NewErrorResponse(ce.HttpCodeForDaoError(err), "Error setting status for snapshotted repository", err.Error())
+	}
```
Comment (jlsherrill) [inline pkg/handler/repositories.go:515]: currently the status is really just tied to introspection.  We need to figure this out longer term though.  I'll open a card for this, but i'd leave out the Update of status

**Comment by jlsherrill** [comment]
Two small changes, and then this is good.

File: pkg/handler/repositories.go (hunk 2 above)
Comment (Andrewgdewar) [reply pkg/handler/repositories.go]: Derp, yeah I know this, that's what I get for pattern following!

**Review by jlsherrill — APPROVED** [review]

**Comment by swadeley** [comment]
Hi @Andrewgdewar , rebase please.

**Comment by swadeley** [comment]
Hi @Andrewgdewar 

I updated API docs.

I created a repo with `'snapshot': False,`

`In [3]: app.content_sources.rest_client.repositories_api.create_repository(repo)
2023-12-04 14:18:23.975 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=VSSNbDLEEygNcEyXnpcCRzldOJoZcboL, params=[]
Out[3]: 
{'account_id': '0369233',
 'content_type': 'rpm',
 'distribution_arch': 'any',
 'distribution_versions': ['any'],
 'failed_introspections_count': 0,
 'gpg_key': '',
 'last_introspection_error': '',
 'last_introspection_time': '',
 'last_success_introspection_time': '',
 'last_update_introspection_time': '',
 'metadata_verification': False,
 'name': 'fedoratest-snapshot-true',
 'org_id': '3340851',
 'origin': 'external',
 'package_count': 0,
 'snapshot': False,
 'status': 'Pending',
 'url': 'https://stephenw.fedorapeople.org/multirepos/5/repo06/',
 'uuid': 'ca7faa4a-eab8-4fc4-80f4-d46e4c97fc42'}`

I created a snapshot:

`In [7]: app.content_sources.rest_client.repositories_api.create_snapshot('ca7faa4a-eab8-4fc4-80f4-d46e4c97fc42')
2023-12-04 14:23:38.083 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=WyKXYVhdvugmInmxJKAtAcAVZrubiCfs, params=[]

In [8]: app.content_sources.rest_client.repositories_api.get_repository('ca7faa4a-eab8-4fc4-80f4-d46e4c97fc42')
2023-12-04 14:24:16.586 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=yIaOKulwjvcCShJyfiTbpIaOsCRmUhmN, params=[]
Out[8]: 
{'account_id': '0369233',
 'content_type': 'rpm',
 'distribution_arch': 'any',
 'distribution_versions': ['any'],
 'failed_introspections_count': 0,
 'gpg_key': '',
 'last_introspection_error': '',
 'last_introspection_time': '2023-12-04T06:18:24Z',
 'last_snapshot': {'added_counts': {'rpm.package': 1},
                   'content_counts': {'rpm.package': 1},
                   'created_at': '2023-12-04T06:23:54.714368Z',
                   'removed_counts': {},
                   'repository_path': 'ff180fe7/ca7faa4a-eab8-4fc4-80f4-d46e4c97fc42/6f07466d-b03f-4f48-9865-ab164dc0d48b',
                   'url': 'https://env-ephemeral-gbpyh4.apps.c-rh-c-eph.8p0c.p1.openshiftapps.com/pulp/content/ff180fe7/ca7faa4a-eab8-4fc4-80f4-d46e4c97fc42/6f07466d-b03f-4f48-9865-ab164dc0d48b/',
                   'uuid': 'fb3fc5e0-182c-43d4-8bfe-5b719c94f1b3'},
 'last_snapshot_task_uuid': '9b3fa3df-86ee-43de-a3db-70d3befe450f',
 'last_snapshot_uuid': 'fb3fc5e0-182c-43d4-8bfe-5b719c94f1b3',
 'last_success_introspection_time': '2023-12-04T06:18:24Z',
 'last_update_introspection_time': '2023-12-04T06:18:24Z',
 'metadata_verification': False,
 'name': 'fedoratest-snapshot-true',
 'org_id': '3340851',
 'origin': 'external',
 'package_count': 1,
 'snapshot': False,
 'status': 'Valid',
 'url': 'https://stephenw.fedorapeople.org/multirepos/5/repo06/',
 'uuid': 'ca7faa4a-eab8-4fc4-80f4-d46e4c97fc42'}

In [9]: 
`

When i check the UI, I see it says "Last snapshot 2 minutes ago", but when I check the kebab menu it says "No snapshots yet".  You could say that is out of scope and fix that later? A front-end ch ange?

**Comment by Andrewgdewar** [comment]
> When i check the UI, I see it says "Last snapshot 2 minutes ago", but when I check the kebab menu it says "No snapshots yet". You could say that is out of scope and fix that later? A front-end change?

Hmmm the described issue sounds almost like the snapshot was taken but unsuccessful, if the last_snapshot: variable exists and is not empty, the snapshot list dropdown item should be populated. 
As the last snapshot time stamp is within the "last_snapshot", I'm not sure how this could happen.. hmm will attempt to recreate.

**Pull request merged**